src/lib/comms/sol-netctl-impl-connman.c

```diff
@@ -0,0 +1,1171 @@
+        sd_bus_slot_unref(_ctx.state_slot);
+    _ctx.manager_slot =
+        sd_bus_slot_unref(_ctx.manager_slot);
+    _ctx.service_slot =
+        sd_bus_slot_unref(_ctx.service_slot);
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&_ctx.service_vector, service, id)
+        _free_connman_service(service);
+
+    sol_ptr_vector_clear(&_ctx.service_vector);
+    sol_monitors_clear(&_ctx.service_ms);
+    sol_monitors_clear(&_ctx.manager_ms);
+    sol_monitors_clear(&_ctx.error_ms);
+
+    _ctx.connman_state = SOL_NETCTL_STATE_UNKNOWN;
+}
+
+static int
+_match_properties_changed(sd_bus_message *m, void *userdata,
+    sd_bus_error *ret_error)
+{
+    const char *interface;
+    int r;
+
+    SOL_NULL_CHECK(_ctx.properties_changed, -EINVAL);
+
+    interface = sd_bus_message_get_interface(m);
+    SOL_NULL_CHECK(interface, -EINVAL);
+
+    if (!strstartswith(interface, "net.connman."))
+        return -EINVAL;
+
+    if (sd_bus_message_is_signal(m, "net.connman.Manager",
+        "ServicesChanged")) {
+        if (_ctx.service_slot)
+            return -EINVAL;
+
+        services_list_changed(m);
+    } else if (sd_bus_message_is_signal(m, "net.connman.Manager",
+        "PropertyChanged")) {
+        char *str = NULL;
+
+        if (_ctx.manager_slot)
+            return -EINVAL;
+
+        r = sd_bus_message_read_basic(m, SD_BUS_TYPE_STRING, &str);
+        SOL_INT_CHECK(r, < 0, r);
+        if (streq(str, "State")) {
+            r = get_manager_properties(m);
+            SOL_INT_CHECK(r, < 0, r);
+            call_manager_monitor_callback();
+        } else {
+            SOL_DBG("Ignored changed property: %s", str);
+            r = sd_bus_message_skip(m, "v");
+            SOL_INT_CHECK(r, < 0, r);
+        }
+    }
+
+    return 0;
+}
+
+static int
+dbus_service_add_monitor(
+    sol_netctl_service_monitor_cb cb, const void *data)
+{
+    int r;
+    struct sol_monitors_entry *e;
+    char matchstr[] = "type='signal',interface='net.connman.Manager'";
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    e = sol_monitors_append(&_ctx.service_ms, (sol_monitors_cb_t)cb, data);
+    SOL_NULL_CHECK(e, -ENOMEM);
+
+    if (_ctx.properties_changed)
+        goto end;
+
+    r = sd_bus_add_match(bus, &_ctx.properties_changed, matchstr,
+        _match_properties_changed, NULL);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+end:
+    if (sol_monitors_count(&_ctx.service_ms) == 1)
+        return 0;
+    else
+        return 1;
+}
+
+static int
+dbus_service_del_monitor(
+    sol_netctl_service_monitor_cb cb, const void *data)
+{
+    int r;
+
+    r = sol_monitors_find(&_ctx.service_ms, (sol_monitors_cb_t)cb, data);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return sol_monitors_del(&_ctx.service_ms, r);
+}
+
+SOL_API int
+sol_netctl_add_service_monitor(
+    sol_netctl_service_monitor_cb cb, const void *data)
+{
+    int r;
+
+    r = sol_netctl_init_lazy();
+    SOL_INT_CHECK_GOTO(r, < 0, fail_init);
```

Comment:
The same here, if the `init()` fails, do you really need call the `shutdown()`  ?